torchreid/engine/engine.py

```diff
@@ -590,19 +586,39 @@ def test(
             ``extract_features()`` and ``parse_data_for_eval()`` (most of the time),
             but not a must. Please refer to the source code for more details.
         """
-        if self.use_ema_decay and not lr_finder and not test_before_train:
-            for ema in self.ema_wrapped_models:
-                ema.apply_shadow()
+
         self.set_model_mode('eval')
         targets = list(self.test_loader.keys())
         top1, mAP, top5 = [None]*3
         cur_top1, cur_mAP, cur_top5 = [None]*3
+        ema_top1, ema_mAP, ema_top5 = [0]*3
+        should_save_ema_model = False
+        if self.use_ema_decay and not lr_finder and not test_only:
+            for dataset_name in targets:
+                domain = 'source' if dataset_name in self.datamanager.sources else 'target'
```

Comment:
Seems like you could move this to the next loop by datasets instead of creating a new one